### PR TITLE
[WIP] Improve encryption transformer lookup performance

### PR DESCRIPTION
**Pull request by dinhxuanvu**
#### What type of PR is this?
/kind cleanup

#### What this PR does / why we need it:
The current implementation of the transformer lookup for encryption is to loop through the entire list of transformers. For the case, multiple keys with multiple encryption types, it can be inefficient.

The list of transformers can be constructed as a map with encryption prefix as a key to improve lookup time while maintaining the input order via a slice of transformers as map value.

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?
No

```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/storage/value/transformer.go [hunk 1]
```diff
@@ -164,3 +189,15 @@ func (t *prefixTransformers) TransformToStorage(ctx context.Context, data []byte
 	prefixedData = append(prefixedData, result...)
 	return prefixedData, nil
 }
+
+// getDataPrefix returns the encryption prefix
+// Format: k8s:enc:<encryption-type>:<encryption-version>:
+// For example: k8s:enc:kms:v2:
+func getDataPrefix(data []byte) string {
```
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/storage/value/transformer.go:196]: This is incorrect, `prefixTransformers` need not be used at only the root, for example see the code in `aesPrefixTransformer`.

**Review by enj — CHANGES_REQUESTED** [review]
Hi @dinhxuanvu, thanks for the PR.  However, this change is neither necessary nor correct.  Any performance gains here would be theoretical, and make an already difficult part of the code harder to read.  We will keep the existing slice based approach because it is easy to read, and the cost of iterating over a slice or two is dwarfed by the cost of the actual crypto and encoding/decoding required.

These changes break the AES transformer which is easy to observe via:

```
make test-integration WHAT=./test/integration/controlplane/transformation KUBE_TEST_ARGS="-v"
```


In the future, please discuss any changes you want to make in the [#sig-auth-kms-dev](https://kubernetes.slack.com/archives/C03035EH4VB) channel before starting the implementation.

/hold

**Comment by dims** [comment]
This work-in-progress PR needs a rebase. please rebase if this PR is still needed for the upcoming release.

**Pull request closed without merging**